cql3: Warn when creating RF-rack-invalid keyspace

dawmd opened the pull request:
Although RF-rack-valid keyspaces are not universally enforced
yet (they're governed by the configuration option
`rf_rack_valid_keyspaces`), we'd like to encourage the user to
abide by the restriction.

To that end, we're introducing a warning when creating or
altering a keyspace. If the configuration option is disabled,
but the user is trying to create an RF-rack-invalid keyspace,
they'll receive a warning.

If the option is turned off, we will also log all of the
RF-rack-invalid keyspaces at start-up.

We provide validation tests.

Fixes scylladb/scylladb#23330

Backport: we'd like to encourage the user to abide by the restriction
even when they don't enforce it to make it easier in the future to
adjust the schema when there's no way to disable it anymore. Because
of that, we'd like to backport it to all relevant versions, starting with 2025.1.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18057/) | :white_check_mark: | sif |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18057/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18057/) | :x: | spider2.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18057/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18057/) | :x: | spider5.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18057/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18057/) | :x: | spider9.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5829/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17128/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9041/) | :white_check_mark: |  |

#### Build Details:
- Duration: 20 hr

dawmd commented on test/boost/schema_change_test.cc:
```diff
@@ -1296,4 +1297,126 @@ SEASTAR_TEST_CASE(metadata_id_unchanged) {
     });
 }
 
+SEASTAR_TEST_CASE(test_warn_create_rf_rack_invalid) {
```
They don't. At the moment, the alter test gets stuck, but I wasn't able to debug it yet. The immediate solution would be to move the test to `cqlpy` or `cluster`. I've been reluctant to do that because:

1. The former suite always runs on exactly one node and we don't have a means to modify the underlying topology. I don't think this PR is that important, but if it's possible, I'd like to test it with many different topology layouts.
2. The latter suite would incur a significant overhead since we'd need to set up a cluster.

That was my reasoning. If these changes should get in fast, I can ignore those reasons (since we're just adding a warning...) and move the testing to one of those suites (I'm personally in favour of `cluster`, but either seems good). If it's not a problem to wait a bit/spend more time on it, I can try to debug the alter test next week.

dawmd replied:
Okay, in that case I'll move the tests to the cluster suite.

dawmd replied:
It's pretty convoluted: it's part of the future we're awaiting, which is why we have no access to it when using `run_async`. :( But it works, thank you.

piodul commented on test/cluster/test_multidc.py:
```diff
@@ -464,3 +465,127 @@ async def test_restart_with_prefer_local(request: pytest.FixtureRequest, manager
+        # OK: RF=1 is always accepted.
+        do_create_test(1, 1, "true", True),
+    ]
+    await asyncio.gather(*tasks)
```
I'm not sure why we have to go to such lengths to run those checks in parallel. After all, they only try to create a keyspace and check that they got an error - it should be rather fast, no?

piodul commented on cql3/statements/alter_keyspace_statement.cc:
```diff
@@ -301,9 +300,22 @@ cql3::statements::alter_keyspace_statement::prepare_schema_mutations(query_proce
+                // we'd like to inform the user that the keyspace they're altering will not
+                // satisfy the restriction after the change--but just as a warning.
+                // For more context, see issue: scylladb/scylladb#23330.
+                warnings.push_back(seastar::format(
+                    "The keyspace '{}' you're trying to alter will not satisfy the requirements "
+                    "to be RF-rack-valid after the change, i.e. it uses tablets, but the replication factor in "
+                    "at least one of the data centers does not match the number of racks in that data center. "
+                    "That may result in worse availability. Consider updating the replication factors "
+                    "to satisfy that.",
+                    _name));
```
Please apply this suggestion. I think it's rather important to be very explicit about the fact that we will start to enforce it in a future version.

dawmd commented:
Version 3:
* Extended the warning.

dawmd commented:
Version 4:
* Set `max_schema_agreement_wait` to 0, which reduced the time required by the test to about 9 seconds. Got rid of `asyncio.gather` as it didn't bring that much benefit anymore.

tzach commented:
I think this warning is a good idea, but will have limited impact. 
We will still need to handle cases when it is ignored, or violated prior to the introducing the warning.

I suggest a simplify, shorter version: 
"After this change the keyspace RF will not match the number of Racks in the DataCenter, which might reduce availability in case of a failure.  For more context, see: ..."

dawmd commented:
Version 5:
* Simplified warnings when creating/altering a keyspace, as suggested by Tzach.
* Implemented logging RF-rack-invalid keyspaces at start-up if the configuration option is disabled.

piodul commented on replica/database.cc:
```diff
@@ -3363,11 +3364,29 @@ database::on_effective_service_levels_cache_reloaded() {
+    if (invalid_keyspaces.size() == 0) {
+        dblog.info("All keyspaces are RF-rack-valid");
+    } else {
+        dblog.warn("The following keyspaces are not RF-rack-valid: {}", invalid_keyspaces
+                | std::views::transform(cql3::util::maybe_quote)
+                | std::views::join_with(std::string_view(", "))
+                | std::ranges::to<std::string>());
```
Please add a more comprehensive message here. Like in the case of create/alter, there should be a short explanation and, ideally, a link to the documentation.

piodul reviewed (commented):
Only one comment from me, otherwise it looks good.

piodul commented on replica/database.cc:
```diff
@@ -3368,11 +3368,34 @@ database::on_effective_service_levels_cache_reloaded() {
+                "does not match the number of racks in one of the datacenters, and it may reduce "
+                "availability in case of a failure (cf. "
+                "https://docs.scylladb.com/manual/stable/reference/glossary.html#term-RF-rack-valid-keyspace). "
+                "Those keyspaces are: {}", ks_list);
```
:+1: for listing the problematic keyspaces

piodul reviewed (approved):
LGTM

@tzach please give your blessing (or comments), I'll wait with merging

tzach commented:
> The reason why we use the term "RF-rack-valid" is because it's a term defined in our documentation

OK, lets keep it

dawmd commented:
Version 7.1:
* Changed the logs.
